AppController/djinn.rb

```diff
@@ -2506,208 +2496,116 @@ def is_app_running(appname, secret)
-  # assigned the roles given to us by the caller.
-  # Args:
-  #   ips_to_roles: A Hash that maps machines to the roles that should be
-  #     started on them. As we have not yet spawned the machines, we do not
-  #     have IP addresses for them, so any unique identifier can be used in
-  #     lieu of IP addresses.
-  # Returns:
-  #   An Array of Strings, where each String contains information about the
-  #     public IP address, private IP address, and roles that the new machines
-  #     have taken on.
-  def start_new_roles_on_nodes_in_cloud(ips_to_roles)
-    Djinn.log_info("Starting new roles in cloud with following info: " +
-      "#{ips_to_roles.inspect}")
-
-    num_of_vms = ips_to_roles.keys.length
-    roles = ips_to_roles.values
-    disks = Array.new(num_of_vms, nil)  # no persistent disks
-    Djinn.log_info("Need to spawn up #{num_of_vms} VMs")
-    imc = InfrastructureManagerClient.new(@@secret)
-
-    begin
-      new_nodes_info = imc.spawn_vms(num_of_vms, @options, roles, disks)
-    rescue FailedNodeException, AppScaleException => exception
-      Djinn.log_error("Couldn't spawn #{num_of_vms} VMs with roles #{roles} " +
-        "because: #{exception.message}")
-      return []
-    end
-
-    # initialize them and wait for them to start up
-    Djinn.log_debug("info about new nodes is " +
-      "[#{new_nodes_info.join(', ')}]")
-
-    add_nodes(new_nodes_info)
-    update_hosts_info()
-
-    return new_nodes_info
-  end
-
-
-  # This method takes a list of IP addresses that correspond to machines
-  # with AppScale installed on them, that have passwordless SSH already
-  # set up (presumably by appscale-add-instances). The machines are added
-  # to the currently running AppScale deployment, and are then assigned
-  # the roles given to us by the caller.
-  # Args:
-  #   ips_to_roles: A Hash that maps machines to the roles that should be
-  #     started on them. Machines are uniquely identified by their IP
-  #     address, which is assumed to be reachable from any node in the
-  #     AppScale deployment.
-  # Returns:
-  #   An Array of Strings, where each String contains information about the
-  #     public IP address, private IP address, and roles that the new machines
-  #     have taken on.
-  def start_new_roles_on_nodes_in_xen(ips_to_roles)
-    Djinn.log_info("Starting new roles in virt with following info: " +
-      "#{ips_to_roles.inspect}")
-
-    nodes_info = []
-    ips_to_roles.each { |ip, roles|
-      Djinn.log_info("Will add roles #{roles.join(', ')} to new " +
-        "node at IP address #{ip}")
-      nodes_info << {
-        "public_ip" => ip,
-        "private_ip" => ip,
-        "jobs" => roles,
-        "disk" => nil
+      ips.each { |ip_or_node|
+        begin
+          # Convert (or check) if we have an IP address or we have a node
+          # we need to start, then we add this role to the node.
+          ip = HelperFunctions.convert_fqdn_to_ip(ip_or_node)
```

Comment:
Not really. The idea is that the rescue block is a name that cannot be converted (so we hope that node-x cannot be converted), which means that is a node we need to add by scratch. If we were to force IP address or node-x (instead of IP/hostname or node-x) we could get by without this conversion.